### PR TITLE
New QuantEcon/base

### PR DESCRIPTION
**Pull request by arnavs**
## Assets 

| Environment | Resources | 
| --------------- | -------------- | 
| Linux | `git, wget, pandoc, texlive-latex-base, texlive-xetex, inkscape, ...` | 
| Julia (v1.0 Environment) | `Expectations, Compat, IJulia, Interpolations, Revise, QuantEcon, InstantiateFromURL, Plots, GR, Parameters, Distributions` | 
| Julia (`QELP`) | Whatever's in the latest `https://github.com/QuantEcon/QuantEconLecturePackages`, as well as the last tagged release. | 
| Python | `quantecon, pandas, numba, numpy, scipy, urllib3, matplotlib, linearmodels...` | 

## Overview/Changes

1. Instead of using local `TOML`, we use the `InstantiateFromURL.jl` setup (along with a few base packages in the `v1.0` environment), so that there's only one set of dependencies to maintain. 

2. Instead of basing off of `jupyter/scipy-notebook`, we use the `jupyter/base-notebook` to manage image size. 

3. As per @sglyon, we decided to persist the Julia user depot in a Docker volume. I think the most elegant way to do this is to specify a `VOLUME` statement from inside the Dockerfile, and then give that volume a name at runtime. See below. 

## Usage 

@jlperla I made a few changes; here are the new instructions. Will put these into the lectures soon. These should be followed in order. 

> Install the Docker image. 

Run `docker pull quantecon/base`. 

> Clear any existing `qe-julia` Docker data volume (or other name).

Run `docker volume rm qe-julia`. 

> Create a container. 

Run `docker run --rm -p 8888:8888 -v qe-julia:/home/jovyan/.julia quantecon/base`. 

This is worth unpacking. The `--rm` flag says that Docker should garbage-collect this container (but not the associated data volume) upon exit. (The `-d` flag had some bad side-effects, like suppressing the token for the Jupyter notebook). The `-p` flag is standard. The `-v` flag will either: 

- If the volume doesn't exist, create it, fill it with the `~/.julia` specified by the image, and then proceed to step 2. 
- If the volume does exist, use it in place of the `~./julia` specified by the image. 

> Stop a container. 

Just `ctrl+c` it, or run `docker kill $(docker ps -aq)` in the terminal. 

> Run a container in the local folder. 

This will require specifying two volumes, or: 

```
(Linux) docker run --rm -p 8888:8888 -v qe-julia:/home/jovyan/.julia -v "$PWD":/home/jovyan/local quantecon/base 
(Powershell) docker run --rm -p 8888:8888 -v qe-julia:/home/jovyan/.Julia -v {pwd}:/home/jovyan/local quantecon/base 
```

### CODE REVIEW

**Comment by arnavs** [comment]
@jlperla In order to use this persistently, run `docker create volume quantecon` to create a volume named `quantecon`, and mount it using `docker run -p 8888:8888 -v quantecon:/home/jovyan/work -d quantecon/base`. Everything students save to the `work` directory (including package versions installed on top of the base) will be persisted between containers.

**Comment by arnavs** [comment]
(I'm still making some tweaks, but nothing that changes core functionality or structure).

**Comment by jlperla** [comment]
All sounds great! What do you guys think about preparing a simple shell command and bat file which formalize the commands? Then we could tell people to put it somewhere in their path? The instructions could have it download as a link in some repo (eg static build assets)

**Comment by arnavs** [comment]
Sure, we can do that @jlperla

File: base/Dockerfile [hunk 1]
```diff
@@ -1,55 +1,77 @@
...
+
+# Install Julia packages 
+# Base packages
+RUN julia -e "using Pkg; pkg\"add Expectations Compat IJulia Interpolations Revise QuantEcon InstantiateFromURL Plots GR Parameters Distributions\"; pkg\"build\"; pkg\"precompile\"" 
+# v0.3.1 QELP
+RUN julia -e "using InstantiateFromURL; activate_github(\"QuantEcon/QuantEconLecturePackages\", tag = \"v0.3.1\")"
+# Master QELP 
+RUN julia -e "using InstantiateFromURL; activate_github(\"QuantEcon/QuantEconLecturePackages\")"
+# Pkg3 garbage collection 
+RUN julia -e "using Pkg; pkg\"gc\""
```
Comment (sglyon) [inline base/Dockerfile]: I think that we should combine the steps to install julia packages into one `RUN` command and make sure to have `pkg"gc"` run at the end of this command. 

As it stands right now anything that would be `gc`d by julia's Pkg is already baked into the image by the previous `RUN` commands that install packages.

**Comment by jlperla** [comment]
Are we sure the `gc` is a good idea? I think we should avoid it. "One man's garbage is another man's precompiled package versions" ... Especially with the new Julia package system where you would frequently be using different package versions simultaneously

**Pull request merged**